Sources/Core/Shared/GroupOperation.swift

```diff
@@ -185,22 +248,38 @@ public class GroupOperation: Operation, OperationQueueDelegate {
 
     /**
      The group operation acts as its own queue's delegate. When an operation is added to the queue,
-     assuming that the finishing operation has not started (or finished), and the operation is not
-     the finishing operation itself, then we add the operation as a dependency to the finishing
-     operation.
+     assuming that the finishing operation has not started (or finished), the operation is not
+     the finishing operation itself, and the CanFinishOperation has not yet started,
+     then we add the operation as a dependency to the CanFinishOperation.
 
      The purpose of this is to keep the finishing operation as the last child operation that executes
-     when there are no more operations in the group.
+     when there are no more operations in the group, and to safely handle GroupOperation state changes.
      */
     public func operationQueue(queue: OperationQueue, willAddOperation operation: NSOperation) {
+        guard queue === self.queue else { return }
+
         assert(!finishingOperation.executing, "Cannot add new operations to a group after the group has started to finish.")
         assert(!finishingOperation.finished, "Cannot add new operations to a group after the group has completed.")
 
         if operation !== finishingOperation {
+            let shouldContinue = groupFinishLock.withCriticalScope { () -> Bool in
+                guard !isGroupFinishing else {
+                    assertionFailure("Cannot add new operations to a group after the group has started to finish.")
+                    return false
+                }
+                dispatch_group_enter(isAddingOperationsGroup)
+                return true
+            }
+            
+            guard shouldContinue else { return }
 
             willAddChildOperationObservers.forEach { $0.groupOperation(self, willAddChildOperation: operation) }
 
-            finishingOperation.addDependency(operation)
+            canFinishOperation.addDependency(operation)
+            
+            groupFinishLock.withCriticalScope {
+                dispatch_group_leave(isAddingOperationsGroup)
+            }
```

Comment:
Discussed with @swiftlyfalling - in essence this is because `dispatch_group_enter/leave`'s locking mechanism is noticeably slower than `NSRecursiveLock`.